File: FromSlides/C/18_blur_parallel_inner.c
#include <stddef.h>
#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include <unistd.h>
#include <omp.h>

#define NX 1002
#define NY 1002

void blur(int *image, size_t szx, size_t szy, size_t iters){

  int *temp = malloc(sizeof(int) * szx * szy);
  for (size_t i = 0; i< NX*NY; ++i) temp[i]=image[i];

  for (size_t iit = 0; iit < iters; ++iit){
    for (size_t ix = 1; ix< szx-1; ++ix){
#pragma omp parallel for
      for (size_t iy = 1; iy< szy-1; ++iy){
        temp[iy + ix * szy] = (int)(0.25 * (float)(image[iy + (ix+1) * szy] +
            image[iy + (ix-1) * szy] + image[(iy-1) + ix * szy] +
            image[(iy+1) + ix * szy]) + 0.5);
      }
    }
    for (size_t i = 0; i < (szx * szy); ++i){
      image[i] = temp[i];
    }
  }
  free(temp);
}

int main(){

  int image[(NX)*(NY)];
  struct timespec t1, t2;
  float dtime;

  for (size_t i = 0; i< NX*NY; ++i) image[i]=5;

  printf("OpenMP code running on %i threads\n",omp_get_max_threads());

  clock_gettime(CLOCK_REALTIME, &t1);
  blur(image,NX,NY, 10000);
  clock_gettime(CLOCK_REALTIME, &t2);
  dtime = (float)(t2.tv_sec - t1.tv_sec) + ((float)(t2.tv_nsec - t1.tv_nsec)
      /1.0e9);
  printf("Time taken was %f seconds\n",dtime);
  printf("Arbitrary value from image %i\n",image[100]);
  printf("Arbitrary value printed to avoid compiler optimising the blur out\n");

}

File: FromSlides/Fortran/18_blur_parallel_inner.f90
MODULE blurmod
  IMPLICIT NONE
  CONTAINS

  SUBROUTINE blur(image, its)

    INTEGER, DIMENSION(0:,0:), INTENT(INOUT) :: image
    INTEGER, INTENT(IN) :: its
    INTEGER :: ix, iy, iit, UB1, UB2
    INTEGER, DIMENSION(:,:), ALLOCATABLE :: temp

    ALLOCATE(temp, SOURCE = image)

    UB1 = UBOUND(image,2) - 1
    UB2 = UBOUND(image,1) - 1

!$OMP PARALLEL PRIVATE(iit, iy)
    DO iit = 1, its
      DO iy = 1, UB1
!$OMP DO
        DO ix = 1, UB2
          temp(ix,iy) = INT(0.25 * REAL(image(ix+1,iy) + image(ix-1,iy) &
              + image(ix,iy+1) + image(ix,iy-1)))
        END DO
!$OMP END DO
      END DO
!$OMP WORKSHARE
      image = temp
!$OMP END WORKSHARE
    END DO
!$OMP END PARALLEL

  END SUBROUTINE blur
END MODULE blurmod


PROGRAM omptest

  USE blurmod
  USE omp_lib
  IMPLICIT NONE
  INTEGER, DIMENSION(0:1001, 0:1001) ::image
  INTEGER :: t1, t2
  REAL :: tick

  image = 5
  PRINT '(A,I0,A)','OpenMP code running on ', omp_get_max_threads(), ' threads'
  CALL SYSTEM_CLOCK(COUNT_RATE = tick)
  CALL SYSTEM_CLOCK(COUNT=t1)
  CALL blur(image, 10000)
  CALL SYSTEM_CLOCK(COUNT=t2)
  PRINT '(A,F5.2,A)','Time taken was ',REAL(t2-t1)/tick, ' seconds'
  PRINT '(A,I0)', 'Maximum of image was ', MAXVAL(image)
  PRINT '(A)', 'Maximim value printed to avoid compiler optimising the blur out'

END PROGRAM omptest
